addons/mail/static/src/models/thread/thread.js

```diff
@@ -452,18 +459,32 @@ function factory(dependencies) {
 
         /**
          * Mark the specified conversation as read/seen.
+         * @param {mail.message} [message] the message to be considered as last seen, by default
+         * last message of the thread.
          */
-        async markAsSeen() {
-            if (this.message_unread_counter === 0) {
+        markAsSeen(message) {
+            if (!message) {
+                message = this.lastMessage;
+            }
+            if (!message) {
                 return;
             }
+            if (this.pendingSeenMessageId && message.id <= this.pendingSeenMessageId) {
+                return;
+            }
+            if (this.seen_message_id && message.id <= this.seen_message_id) {
+                return;
+            }
+            this.update({ pendingSeenMessageId: message.id });
             if (this.model === 'mail.channel') {
-                const seen_message_id = await this.async(() => this.env.services.rpc({
+                this.env.services.rpc({
                     model: 'mail.channel',
                     method: 'channel_seen',
-                    args: [[this.id]]
-                }, { shadow: true }));
-                this.update({ seen_message_id });
+                    args: [[this.id]],
+                    kwargs: {
+                        last_message_id: message.id,
+                    },
+                }, { shadow: true });
             }
             this.update({ message_unread_counter: 0 });
```

Comment:
This is dangerous to keep, because it can easily become unsync. (eg. received a new message while marking and older message as read)
Probably should be computed from an initial "server value" + the difference between last seen message id and new known messages.

And this must definitely be done outside of this flux where the rpc is called, because all window/browser syncing with the bus will not execute this rpc code.

Reply:
I'm not sure to understand ...

Reply:
Reset of `message_unread_counter` to 0 here is unreliable and should not be flux specific.

This will probably be a complex change but this line cannot stay here.

Reply:
This could be done in the handling of the  `channel_seen` event from the bus couldn't it ?

Reply:
Yes, it would already be better in the handler, this should be its place.

But just moving it there, while good, is still not enough. Resetting to 0 without further precautions during an async process is bad.

Example: If a new message arrives between the request of mark as read and the response, the counter will be incorrectly set to 0, while it should be 1.

The server should communicate which is the last message id that was marked as read, and the client should reset the counter to "0 + all unread message with id > last id".

Reply:
Mmh... what do you do if the client is not aware of all unread messages ? (I'm not sure it can happen)

Reply:
Agree, that's a problem, that's why I said this would be complex...
Ideally the server should send the unread value, and it should always be together with the id of the last message that was taken into account to compute it.